target/linux/lantiq/dts/RS353.dts

```diff
@@ -0,0 +1,318 @@
+/dts-v1/;
+
+/include/ "vr9.dtsi"
+
+/ {
+	model = "RS353 - Bintec Elmeg RS353";
+
+	chosen {
+		bootargs = "console=ttyLTQ0,115200 init=/etc/preinit";
+		leds {
+                        running = &status;
+                        boot = &status;
+			wifi = &wlan;
+			usb2 = &usb; /* Front port is mapped to device 2 */
+			dsl = &dsl;
+                };
+
+	};
+
+	memory@0 {
+		reg = <0x0 0x8000000>;
+	};
+
+	sram@1F000000 {
+		cgu0: cgu@103000 {
+                        compatible = "lantiq,cgu-xway";
+                        reg = <0x103000 0x1000>;
+                        #clock-cells = <1>;
+			lantiq,phy-clk-src = <0x2>;
+		};
+	};
+
+	fpi@10000000 {
+		localbus@0 {
+			nor-boot@0 {
+				compatible = "lantiq,nor";
+				bank-width = <2>;
+				reg = <0 0x0 0x2000000>;
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				partitions {
+					compatible = "fixed-partitions";
+					#address-cells = <1>;
+					#size-cells = <1>;
+
+					boardconfig: partition@0 {
```

Comment:
> What do you think about adding an option in the dts files which enables the possibility to disable the endianness workaround ? At the beginning of lantiq-flash driver, you can see this:

What I see so far, the register in question is at 0xbe105360 and the last bit in this register does make the difference.

```
1000000000000010001000110100111 (0x400111a7)
0000000000000010001000110100111 (0x000111a7)
```

Would you please check the kernel/u-boot code for the meaning of this bit or the name of the register! Maybe it is enough to change to code to apply the NOR endianess workaround based on whether this bit is set or not.

It can be done via device tree as well. But the device tree should describe the hardware and should not apply workarounds because of bootloader bugs or similar. It worked for my lantiq-phy-clk since which (hardware)-clock to use describes the hardware.

>Regarding the flash layout, I'll try to keep the bootmonitor. I'll make a post-install (or first-boot) script that updates the mtd with new size and crc.

Keeping the bootmonitor seams to me a good decision. Should make the revert to the bintec firmware really easy.

> The preloader contains only an asm jump to the address 100e0040 (firmware plus the uImage header size offset 0x40).
>```
> define Device/RS353
>  DEVICE_PROFILE := RS353
>  DEVICE_TITLE := Bintec RS353
>  KERNEL := kernel-bin | append-dtb | lzma | lzma-loader | uImage lzma
>  IMAGE_SIZE := 31232k
>  DEVICE_PACKAGES := kmod-usb-dwc2 kmod-ath9k wpad-mini kmod-i2c-gpio kmod-swconfig wpad-mini kmod-ledtrig-usbdev
> endef
>TARGET_DEVICES += RS353
>```

Both is really ugly. Please consider adding the bossheader to your image instead of a not required uImage header. It is obvious that it's a hack. If you add the bossheader, you can write an custom mtdsplitter which uses the length values from the bossheader. It shouldn't be that difficult.

Your image build code could be in the end something like this:

```
 define Device/RS353
  DEVICE_PROFILE := RS353
  DEVICE_TITLE := Bintec RS353
  KERNEL := kernel-bin | append-dtb | lzma | lzma-loader | gzip
  IMAGE_SIZE := 31232k
  IMAGE/sysupgrade.bin  := append-kernel | append-rootfs | pad-rootfs | boss-header | check-size $$$$(IMAGE_SIZE)
  DEVICE_PACKAGES := kmod-usb-dwc2 kmod-ath9k wpad-mini kmod-i2c-gpio kmod-swconfig wpad-mini kmod-ledtrig-usbdev
 endef
TARGET_DEVICES += RS353
```

Depending on whether you can cheat the firmware upgrade routines (boot not gzip compress code, write image bigger than size in bossheader to flash), it can be something like:

```
  KERNEL := kernel-bin | append-dtb | lzma | lzma-loader
  IMAGE/sysupgrade.bin  := append-kernel | append-rootfs | boss-header | pad-rootfs  | check-size $$$$(IMAGE_SIZE)
```
```